PyTorchLightning/pytorch-lightning
# Fix tests related to DDP communication hooks

## What does this PR do?
Part of #12373.

Fixes DDP communication hook tests that are not running in our current CI setup.

### Description of the changes

- Removes `DistributedDataParallel.get_ddp_logging_data()` which only works with `torch<1.9`. In newer PyTorch, it's renamed to `DistributedDataParallel._get_ddp_logging_data() -> Dict`.
- Moves assertions to `DDPStrategy.teardown` because `DistributedDataParallel` is not available after `trainer.fit()` runs as it gets removed during its teardown (related to #8490).

### How to verify the changes
Note that the modified tests don't run in this PR. To verify the changes here, please either
- check out [this GPU CI result](https://dev.azure.com/PytorchLightning/pytorch-lightning/_build/results?buildId=69032&view=logs&j=3afc50db-e620-5b81-6016-870a6976ad29&t=d9f671c5-a304-5675-5394-961fd7f98b9b) running in #12373 OR
- run tests locally with PyTorch 1.10+:
  ```shell
  # from the host
  git clone https://github.com/PyTorchLightning/pytorch-lightning.git
  cd pytorch-lightning
  docker pull pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.11  # or torch1.10
  docker run --rm -it --gpus all --ipc host -v $(pwd):/work -w /work pytorchlightning/pytorch_lightning:base-cuda-py3.7-torch1.11  # or torch1.10

  # inside the container
  export PL_RUN_STANDALONE_TESTS=1
  python -m pytest tests/strategies/test_ddp_strategy_with_comm_hook.py::test_ddp_fp16_compress_comm_hook
  python -m pytest tests/strategies/test_ddp_strategy_with_comm_hook.py::test_ddp_sgd_comm_hook
  python -m pytest tests/strategies/test_ddp_strategy_with_comm_hook.py::test_ddp_fp16_compress_wrap_sgd_comm_hook
  python -m pytest tests/strategies/test_ddp_strategy_with_comm_hook.py::test_ddp_post_local_sgd_comm_hook
  ```

### Does your PR introduce any breaking changes? If yes, please list them.
None

## Before submitting

- [x] Was this **discussed/approved** via a GitHub issue? (not for typos and docs)
- [x] Did you read the [contributor guideline](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/.github/CONTRIBUTING.md), **Pull Request** section?
- [x] Did you make sure your **PR does only one thing**, instead of bundling different changes together?
- [n/a] Did you make sure to **update the documentation** with your changes? (if necessary)
- [n/a] Did you write any **new necessary tests**? (not for typos and docs)
- [x] Did you verify new and **existing tests pass** locally with your changes?
- [x] Did you list all the **breaking changes** introduced by this pull request?
- [n/a] Did you **update the [CHANGELOG](https://github.com/PyTorchLightning/pytorch-lightning/blob/master/CHANGELOG.md)**? (not for typos, docs, test updates, or minor internal changes/refactors)



## PR review

Anyone in the community is welcome to review the PR.
Before you start reviewing, make sure you have read the [review guidelines](https://github.com/PyTorchLightning/pytorch-lightning/wiki/Review-guidelines). In short, see the following bullet-list:

- [x] Is this pull request ready for review? (if not, please submit in draft mode)
- [x] Check that all items from **Before submitting** are resolved
- [x] Make sure the title is self-explanatory and the description concisely explains the PR
- [x] Add labels and milestones (and optionally projects) to the PR so it can be classified

## Did you have fun?

Make sure you had fun coding 🙃


cc @borda @akihironitta @justusschock @kaushikb11 @awaelchli @rohitgr7
